core/coreunix/metadata.go

```diff
@@ -1,20 +1,345 @@
+	}
+
+	var chunkSize int64 = chunker.DefaultBlockSize
+	maxlinks := ihelper.DefaultLinksPerBlock
+
+	dagmod, err := mod.NewDagModifierBalanced(ctx, mnode, modifier.dagService, chunker.MetaSplitterGen((chunkSize)), maxlinks, metadataNotExist)
+	if err != nil {
+		return nil, err
+	}
+
+	params := ihelper.DagBuilderParams{
+		Dagserv:    modifier.dagService,
+		CidBuilder: modifier.CidBuilder,
+		ChunkSize:  uint64(chunkSize),
+		Maxlinks:   maxlinks,
+	}
+	dr := []byte(modifier.TokenMetadata)
+	db, err := params.New(chunker.NewMetaSplitter(bytes.NewReader(dr), uint64(chunkSize)))
+	if err != nil {
+		return nil, err
+	}
+
+	mdagmod := mod.NewMetaDagModifierBalanced(dagmod, db)
+
+	// Add the given TokenMetadata to the metadata sub-DAG and return
+	// a new root of the BTFS file DAG.
+	newRoot, err := mdagmod.AddMetadata(nd, dr)
+	if err != nil {
+		return nil, err
+	}
+
+	return newRoot, nil
+}
+
+// RemoveMetaAndPin removes the metadata entries from the DAG
+// of the given `node` root node of a BTFS file.
+func (modifier *MetaModifier) RemoveMetaAndPin(node ipld.Node) (ipld.Node, ipath.Resolved, error) {
+	if modifier.Pin {
+		modifier.unlocker = modifier.gcLocker.PinLock()
+	}
+	defer func() {
+		if modifier.unlocker != nil {
+			modifier.unlocker.Unlock()
+		}
+	}()
+
+	p, err := modifier.removeMetaItemsFromFileNode(node)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	root, err := modifier.updateMfs()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	nd, err := root.GetNode()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if !modifier.Pin {
+		return nd, p, nil
+	}
+	return nd, p, modifier.PinRoot(nd)
+}
+
+func (modifier *MetaModifier) removeMetaItemsFromFileNode(node ipld.Node) (ipath.Resolved, error) {
+	err := modifier.maybePauseForGC()
+	if err != nil {
+		return nil, err
+	}
+
+	dagnode, err := modifier.removeMeta(node)
+	if err != nil {
+		return nil, err
+	}
+
+	// patch it into the root
+	return modifier.addNodeToMfs(dagnode, "")
+}
+
+// Remove the metadata entries keyed by the given `modifier.TokenMetadata`
+// key values from the given `node`.
+func (modifier *MetaModifier) removeMeta(nd ipld.Node) (ipld.Node, error) {
+	// Error cases
+	if nd == nil {
+		return nil, errors.New("Invalid argument: nil node value")
+	}
+
+	// Build Metadata DAG modifier for balanced format
+	// TODO: trickle format
+	ctx, closer := context.WithCancel(context.Background())
+	defer closer()
+
+	children, err := ft.GetChildrenForDagWithMeta(ctx, nd, modifier.dagService)
+	if err != nil {
+		return nil, err
+	} else {
+		if children == nil {
+			return nil, errors.New("no metadata exists for the given node")
+		}
+	}
+	mnode := children.MetaNode
+
+	// Retrieve SuperMeta info to build meta Dag modifier.
+	// Note that SuperMeta may or may not exist.
+	// TODO: unit test to check the cases..
+	b, err := GetMetaDataFromDagRoot(ctx, nd, modifier.dagService)
+	if err != nil {
+		return nil, err
+	}
+	var superMeta ihelper.SuperMeta
```

Comment:
why does removeMeta look at existing superMeta, but addMeta does not look at existing superMeta (if exists)

Reply:
Refined the code blocks.